src/snarks/groth16/mpc_utils.tcc

```diff
@@ -172,52 +239,198 @@ srs_mpc_layer_L1<ppT> mpc_compute_linearcombination(
+        out << h;
+    }
+
+    for (const G1 l : L_g1) {
+        out << l;
+    }
+}
+
+template<typename ppT>
+srs_mpc_layer_C2<ppT> srs_mpc_layer_C2<ppT>::read(std::istream &in)
+{
+    using G1 = libff::G1<ppT>;
+
+    size_t H_size;
+    size_t L_size;
+
+    in.read((char *)&H_size, sizeof(H_size));
+    in.read((char *)&L_size, sizeof(L_size));
+
+    libff::G1<ppT> delta_g1;
+    libff::G2<ppT> delta_g2;
+    libff::G1_vector<ppT> H_g1(H_size);
+    libff::G1_vector<ppT> L_g1(L_size);
+
+    in >> delta_g1;
+
+    in >> delta_g2;
+
+    for (G1 &h : H_g1) {
+        in >> h;
+    }
+
+    for (G1 &l : L_g1) {
+        in >> l;
+    }
+
+    return srs_mpc_layer_C2<ppT>(
+        delta_g1, delta_g2, std::move(H_g1), std::move(L_g1));
+}
+
+template<typename ppT>
+srs_mpc_layer_C2<ppT> mpc_dummy_layer_C2(
+    const srs_mpc_layer_L1<ppT> &layer1,
     const libff::Fr<ppT> &delta,
-    libsnark::r1cs_constraint_system<libff::Fr<ppT>> &&cs,
-    const libsnark::qap_instance<libff::Fr<ppT>> &qap)
+    const size_t num_inputs)
 {
     using Fr = libff::Fr<ppT>;
     using G1 = libff::G1<ppT>;
     using G2 = libff::G2<ppT>;
+    libff::enter_block("call to mpc_dummy_layer2");
 
     const Fr delta_inverse = delta.inverse();
+    // { H_i } = { [ t(x) . x^i / delta ]_1 } i = 0 .. n-2 (n-1 entries)
+    libff::enter_block("computing H_g1");
+    const size_t H_size = layer1.T_tau_powers_g1.size();
+    libff::print_indent();
+    printf("%zu entries\n", H_size);
+    libff::G1_vector<ppT> H_g1(H_size);
+
+#ifdef MULTICORE
+#pragma omp parallel for
+#endif
+    for (size_t i = 0; i < H_size; ++i) {
+        if ((i % 10240) == 0) {
+            putchar('.');
+            fflush(stdout);
+        }
 
-    // { H_i } = { [ t(x) . x^i / delta ]_1 } i = 0 .. n-2
-    libff::G1_vector<ppT> T_tau_powers_over_delta_g1(
-        layer1.T_tau_powers_g1.size());
-    for (size_t i = 0; i < layer1.T_tau_powers_g1.size(); ++i) {
-        T_tau_powers_over_delta_g1[i] =
-            delta_inverse * layer1.T_tau_powers_g1[i];
+        H_g1[i] = delta_inverse * layer1.T_tau_powers_g1[i];
     }
+    putchar('\n');
+    libff::leave_block("computing H_g1");
+
+    // In layer1 output, there should be num_variables+1 entries in
+    // ABC_g1.  Of these:
+    //
+    //  - The first 1+num_inputs entries are used directly in the
+    //    verification key.
+    //
+    //  - The remaining num_variables-num_inputs entries will be
+    //    divided by delta to create layer2.
+    const size_t num_variables = layer1.ABC_g1.size() - 1;
+    const size_t num_L_elements = num_variables - num_inputs;
+    // { L_i } = { [ ABC_i / delta ]_1 }, i = l+1 .. num_variables
+    libff::enter_block("computing L_g1");
+    libff::print_indent();
+    printf("%zu entries\n", num_L_elements);
+    libff::G1_vector<ppT> L_g1(num_L_elements);
+#ifdef MULTICORE
+#pragma omp parallel for
+#endif
+    for (size_t i = 0; i < num_L_elements; ++i) {
+        if ((i % 10240) == 0) {
```

Comment:
What does `10240` represent?

Reply:
This is just a value that gives a reasonable indication of progress in the cases where degree ~2^20 (balance of update interval and the number of '.'s written).  Let me know if you prefer it changed or removed.

Reply:
I am not sure I understand. I'm happy to keen this block it it enhance the CLI and emulates a progress bar, but I didn't understand why `10240`. Either way, I think it's a good idea to add a comment above to raise any doubts and make it easier to understand when we re-open this file in six months.

Reply:
I just removed it in the end.  This is no longer a particularly expensive computation, so not really necessary to give an indication of progres.